### PR TITLE
ExprCommandInfo Enhancements

### PR DESCRIPTION
**Pull request by UnderscoreTud**
### Description
This PR makes ExprCommandInfo default to the current command if no command was passed into the expression. It also fixes a bug where `isSingle` returned the opposite value

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** #5885

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 1]
```diff
@@ -43,108 +47,105 @@
...
-					} else if (command.getClass().getPackage().getName().startsWith("org.spigot")) {
-						result.add("Spigot");
-					} else if (command.getClass().getPackage().getName().startsWith("com.destroystokyo.paper")) {
-						result.add("Paper");
-					}
-				}
-				break;
+	protected String[] get(Event event) {
+		Command[] commands;
+		if (commandName == null) {
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:124]: can we move this `commands` computation logic to it's own private method? that way it'll be easier to follow

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 2]
```diff
@@ -43,108 +47,105 @@
...
+
+	@Nullable
+	private Expression<String> commandName;
 
 	@Override
+	@SuppressWarnings("unchecked")
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		commandName = (Expression<String>) exprs[0];
+		if (commandName == null && !getParser().isCurrentEvent(ScriptCommandEvent.class))
+			return false;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:115]: Maybe provide an error message as well? something like `"You either provide a command or use it within a command section"`

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 3]
```diff
@@ -43,108 +47,105 @@
...
+@Examples({
+	"main name of command \"skript\"",
 	"description of command \"help\"",
 	"label of command \"pl\"",
 	"usage of command \"help\"",
 	"aliases of command \"bukkit:help\"",
 	"permission of command \"/op\"",
 	"command \"op\"'s permission message",
-	"command \"sk\"'s plugin owner"})
+	"command \"sk\"'s plugin owner"
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:58]: Can we provide an example to use within a command section?

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 4]
```diff
@@ -43,108 +47,105 @@
...
+	private InfoType type;
+
+	@Nullable
+	private Expression<String> commandName;
 
 	@Override
+	@SuppressWarnings("unchecked")
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		commandName = (Expression<String>) exprs[0];
+		if (commandName == null && !getParser().isCurrentEvent(ScriptCommandEvent.class))
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:114]: Can we also support `PlayerCommandPreprocessEvent.class`, `ServerCommandEvent.class`? (on command)

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 5]
```diff
@@ -159,6 +163,31 @@ public String toString(@Nullable Event event, boolean debug) {
...
+		if (event instanceof ServerCommandEvent) {
+			commandName = ((ServerCommandEvent) event).getCommand();
+		} else if (event instanceof PlayerCommandPreprocessEvent) {
+			commandName = ((PlayerCommandPreprocessEvent) event).getMessage().substring(1);
+		} else {
+			return null;
+		}
+		commandName = commandName.split(":")[0];
+		Command command = map.getCommand(commandName);
+		return command != null ? new Command[] {command} : null;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:188]: These will never be reached due to L183

Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:188]: What do you mean? L184 is only reached if the previous 2 conditions weren't met, aka the event wasn't `ServerCommandEvent` or a `PlayerCommandEvent`

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:188]: there is an else section that returns null

Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:188]: Yes... But the other conditions don't return, so if one of them passes then the rest of the code is ran

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:188]: Pardon me, I should never review when I am sleepy, just got what you mean 😁

**Review by AyhamAl-Ali — APPROVED** [review]
LGTM

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 6]
```diff
@@ -43,108 +50,106 @@
...
+				return "Bukkit";
+			} else if (command.getClass().getPackage().getName().startsWith("org.spigot")) {
+				return "Spigot";
+			} else if (command.getClass().getPackage().getName().startsWith("com.destroystokyo.paper")) {
+				return "Paper";
+			}
+			return "Unknown";
+		});
+
+		private final Function<Command, String> function;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:96]: nullable annotation needed

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 7]
```diff
@@ -43,108 +50,106 @@
...
+
+		InfoType(Function<Command, String> function) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:98]: nullable annotation needed

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 8]
```diff
@@ -43,108 +50,106 @@
...
-		Skript.registerExpression(ExprCommandInfo.class, String.class, ExpressionType.SIMPLE,
-			"[the] main command [label] of command %strings%", "command %strings%'[s] main command [name]",
-			"[the] description of command %strings%", "command %strings%'[s] description",
-			"[the] label of command %strings%", "command %strings%'[s] label",
-			"[the] usage of command %strings%", "command %strings%'[s] usage",
-			"[(all|the|all [of] the)] aliases of command %strings%", "command %strings%'[s] aliases",
-			"[the] permission of command %strings%", "command %strings%'[s] permission",
-			"[the] permission message of command %strings%", "command %strings%'[s] permission message",
-			"[the] plugin [owner] of command %strings%", "command %strings%'[s] plugin [owner]");
+		Skript.registerExpression(ExprCommandInfo.class, String.class, ExpressionType.PROPERTY,
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:105]: breaking change as your changing the expression loading order to be 2 spots lower than SIMPLE.

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 9]
```diff
@@ -43,108 +50,106 @@
...
+	protected String[] get(Event event) {
+		Command[] commands = getCommands(event);
+		if (type == InfoType.ALIASES) {
+			ArrayList<String> result = new ArrayList<>();
+			for (Command command : commands)
+				result.addAll(getAliases(command));
+			return result.toArray(new String[0]);
 		}
-		return result.toArray(new String[0]);
+		String[] result = new String[commands.length];
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:144]: `commands` can be null here on the `.length` call.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 10]
```diff
@@ -43,108 +50,106 @@
 
 @Name("Command Info")
 @Description("Get information about a command.")
-@Examples({"main name of command \"skript\"",
+@Examples({
+	"main name of command \"skript\"",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:54]: This isn't a valid example, should probably be changed while you're here. ``main command label of command "skript"`` is, though. Also see other comment about mismatched syntax.

File: src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java [hunk 11]
```diff
@@ -43,108 +50,106 @@
...
-			"[the] main command [label] of command %strings%", "command %strings%'[s] main command [name]",
-			"[the] description of command %strings%", "command %strings%'[s] description",
-			"[the] label of command %strings%", "command %strings%'[s] label",
-			"[the] usage of command %strings%", "command %strings%'[s] usage",
-			"[(all|the|all [of] the)] aliases of command %strings%", "command %strings%'[s] aliases",
-			"[the] permission of command %strings%", "command %strings%'[s] permission",
-			"[the] permission message of command %strings%", "command %strings%'[s] permission message",
-			"[the] plugin [owner] of command %strings%", "command %strings%'[s] plugin [owner]");
+		Skript.registerExpression(ExprCommandInfo.class, String.class, ExpressionType.PROPERTY,
+			"[the] main command [label] [of [[the] command[s] %-strings%]]", "command[s] %strings%'[s] main command [name]",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprCommandInfo.java:106]: The syntax here is mismatched. The first uses `main command [label]` and the second uses `main command [name]`. This should be corrected to allow both to work on both instances. 
Also this syntax is just ugly (main command of command "command"), but it's up to you if you have better ideas for it. May not be worth a breaking change.

**Review by sovdeeth — APPROVED** [review]

**Review by TheLimeGlass — APPROVED** [review]

**Pull request merged**